script/install/debootstrap.bash

```diff
@@ -171,5 +171,48 @@ echo "$chroot_cmd update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9
 chroot "$ISOLATE_ROOT" update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-9 75
 # gcc 9 done
 
-# let user know that chroot installs are finished
+[ -z "$TRAVIS" ] && bash script/confirm.bash 'Install JavaScript V8 (submissions in V8 will fail without this!)' && {
+  # JavaScript V8 engine compile and install
+  echo "Preparing to compile and install the JavaScript V8 engine."
+  echo
+  echo "If this fails, visit https://v8.dev/docs/build and follow the instructions."
+  echo "This may take a while ..."
+  echo
+  set -x
+
+  : Making temporary directory
+  mkdir ~/v8-build
+  cd ~/v8-build
```

Comment:
Is there a reason to build V8 outside chroot? I think it would be cleaner to to keep everything related to the judging environment inside chroot (specifically, `./build/install-build-deps.sh` installs packages system-wide).

Running a sequence of commands inside chroot can be painful (e.g. `cd` doesn't persist across commands), but there are some approaches to make it easier:
 - Run bash in chroot and redirect the commands from a file (`chroot "$ISOLATE_ROOT" bash < install-v8.sh`) or from a [here-document](https://en.wikipedia.org/wiki/Here_document#Unix_shells)
 - [Export a function](https://stackoverflow.com/a/30792189)

Reply:
We had issues with `git` I believe

Reply:
I'm currently trying to get V8 building inside chroot using the first method (see 49feb87) but am getting errors. Any ideas?

```
+ ./build/install-build-deps.sh
sudo: unable to resolve host holmes-nztrain2
sudo: unable to resolve host holmes-nztrain2
Hit:1 http://nz.archive.ubuntu.com/ubuntu xenial InRelease
Hit:2 http://nz.archive.ubuntu.com/ubuntu xenial-updates InRelease                                         
Get:3 http://nz.archive.ubuntu.com/ubuntu xenial/multiverse i386 Packages [140 kB]                                                   
Get:4 http://nz.archive.ubuntu.com/ubuntu xenial-updates/multiverse i386 Packages [15.9 kB]                                          
Hit:5 http://archive.ubuntu.com/ubuntu xenial InRelease                                                                              
Get:6 http://security.ubuntu.com/ubuntu xenial-security InRelease [109 kB]                    
Hit:7 http://ppa.launchpad.net/brightbox/ruby-ng/ubuntu xenial InRelease
Get:8 http://archive.ubuntu.com/ubuntu xenial-updates InRelease [109 kB]                                                             
Hit:9 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu xenial InRelease                                              
Hit:10 http://ppa.launchpad.net/openjdk-r/ppa/ubuntu xenial InRelease                                                                
Get:11 http://archive.ubuntu.com/ubuntu xenial/main i386 Packages [1196 kB]                                                
Get:12 http://security.ubuntu.com/ubuntu xenial-security/main i386 Packages [631 kB]                      
Hit:13 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu xenial InRelease                                    
Get:14 http://ppa.launchpad.net/brightbox/ruby-ng/ubuntu xenial/main i386 Packages [8532 B]                                          
Get:15 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu xenial/main i386 Packages [31.3 kB]                                            
Get:16 http://ppa.launchpad.net/openjdk-r/ppa/ubuntu xenial/main i386 Packages [11.8 kB]                                             
Get:17 http://security.ubuntu.com/ubuntu xenial-security/restricted i386 Packages [7224 B]                                           
Get:18 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu xenial/main i386 Packages [85.7 kB]
Get:19 http://archive.ubuntu.com/ubuntu xenial/universe i386 Packages [7512 kB]                                                      
Get:20 http://archive.ubuntu.com/ubuntu xenial-updates/main i386 Packages [892 kB]                                                   
Get:21 http://archive.ubuntu.com/ubuntu xenial-updates/restricted i386 Packages [7580 B]                                             
Fetched 10.8 MB in 41s (259 kB/s)                                                                                                    
Reading package lists... Done
Building apt package list.
Skipping debugging symbols.
Including 32-bit libraries.
Including ARM cross toolchain.
Including NaCl, NaCl toolchain, NaCl ports dependencies.
Finding missing packages...
Packages required:  libasound2:i386 libcap2:i386 libelf-dev:i386 libfontconfig1:i386 libglib2.0-0:i386 libgpm2:i386 libncurses5:i386 libnss3:i386 libpango1.0-0:i386 libpci3:i386 libssl-dev:i386 libssl1.0.0:i386 libtinfo-dev:i386 libudev1:i386 libuuid1:i386 libx11-xcb1:i386 libxcomposite1:i386 libxcursor1:i386 libxdamage1:i386 libxi6:i386 libxrandr2:i386 libxss1:i386 libxtst6:i386 linux-libc-dev:i386 ant apache2-bin autoconf binutils binutils-aarch64-linux-gnu binutils-arm-linux-gnueabihf binutils-mips64el-linux-gnuabi64 binutils-mipsel-linux-gnu bison bzip2 cdbs cmake curl dbus-x11 devscripts dpkg-dev elfutils fakeroot flex g++-5-multilib g++-5-multilib-arm-linux-gnueabihf g++-arm-linux-gnueabihf g++-mingw-w64-i686 gawk gcc-5-multilib-arm-linux-gnueabihf gcc-arm-linux-gnueabihf git-core gperf intltool lib32gcc1 lib32ncurses5-dev lib32stdc++6 lib32z1-dev libapache2-mod-php7.0 libappindicator3-1 libappindicator3-dev libasound2 libasound2-dev libatk1.0-0 libatspi2.0-0 libatspi2.0-dev libav-tools libbluetooth-dev libbrlapi-dev libbrlapi0.6 libbz2-1.0 libbz2-dev libc6 libc6-dev libc6-dev-armhf-cross libc6-i386 libcairo2 libcairo2-dev libcap-dev libcap2 libcups2 libcups2-dev libcurl4-gnutls-dev libdrm-dev libdrm2 libelf-dev libevdev-dev libevdev2 libexpat1 libffi-dev libffi6 libfontconfig1 libfreetype6 libgbm-dev libgbm1 libglib2.0-0 libglib2.0-dev libglu1-mesa-dev libgnome-keyring-dev libgnome-keyring0 libgtk-3-0 libgtk-3-dev libinput-dev libinput10 libjpeg-dev libkrb5-dev libnspr4 libnspr4-dev libnss3 libnss3-dev libpam0g libpam0g-dev libpango1.0-0 libpci-dev libpci3 libpcre3 libpixman-1-0 libpng16-16 libpulse-dev libpulse0 libsctp-dev libspeechd-dev libspeechd2 libsqlite3-0 libsqlite3-dev libssl-dev libstdc++6 libtinfo-dev libtool libudev-dev libudev1 libuuid1 libvulkan-dev libvulkan1 libwayland-egl1-mesa libwww-perl libx11-6 libx11-xcb1 libxau6 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxdmcp6 libxext6 libxfixes3 libxi6 libxinerama1 libxkbcommon-dev libxrandr2 libxrender1 libxslt1-dev libxss-dev libxt-dev libxtst-dev libxtst6 linux-libc-dev-armhf-cross locales openbox p7zip patch perl php7.0-cgi pkg-config python python-cherrypy3 python-crypto python-dev python-numpy python-opencv python-openssl python-psutil python-yaml realpath rpm ruby subversion texinfo uuid-dev wdiff x11-utils xcompmgr xsltproc xutils-dev xvfb xz-utils zip zlib1g

E: Unable to correct problems, you have held broken packages.
The following command failed:  apt-get --just-print install libasound2:i386 libcap2:i386 libelf-dev:i386 libfontconfig1:i386 libglib2.0-0:i386 libgpm2:i386 libncurses5:i386 libnss3:i386 libpango1.0-0:i386 libpci3:i386 libssl-dev:i386 libssl1.0.0:i386 libtinfo-dev:i386 libudev1:i386 libuuid1:i386 libx11-xcb1:i386 libxcomposite1:i386 libxcursor1:i386 libxdamage1:i386 libxi6:i386 libxrandr2:i386 libxss1:i386 libxtst6:i386 linux-libc-dev:i386 ant apache2-bin autoconf binutils binutils-aarch64-linux-gnu binutils-arm-linux-gnueabihf binutils-mips64el-linux-gnuabi64 binutils-mipsel-linux-gnu bison bzip2 cdbs cmake curl dbus-x11 devscripts dpkg-dev elfutils fakeroot flex g++-5-multilib g++-5-multilib-arm-linux-gnueabihf g++-arm-linux-gnueabihf g++-mingw-w64-i686 gawk gcc-5-multilib-arm-linux-gnueabihf gcc-arm-linux-gnueabihf git-core gperf intltool lib32gcc1 lib32ncurses5-dev lib32stdc++6 lib32z1-dev libapache2-mod-php7.0 libappindicator3-1 libappindicator3-dev libasound2 libasound2-dev libatk1.0-0 libatspi2.0-0 libatspi2.0-dev libav-tools libbluetooth-dev libbrlapi-dev libbrlapi0.6 libbz2-1.0 libbz2-dev libc6 libc6-dev libc6-dev-armhf-cross libc6-i386 libcairo2 libcairo2-dev libcap-dev libcap2 libcups2 libcups2-dev libcurl4-gnutls-dev libdrm-dev libdrm2 libelf-dev libevdev-dev libevdev2 libexpat1 libffi-dev libffi6 libfontconfig1 libfreetype6 libgbm-dev libgbm1 libglib2.0-0 libglib2.0-dev libglu1-mesa-dev libgnome-keyring-dev libgnome-keyring0 libgtk-3-0 libgtk-3-dev libinput-dev libinput10 libjpeg-dev libkrb5-dev libnspr4 libnspr4-dev libnss3 libnss3-dev libpam0g libpam0g-dev libpango1.0-0 libpci-dev libpci3 libpcre3 libpixman-1-0 libpng16-16 libpulse-dev libpulse0 libsctp-dev libspeechd-dev libspeechd2 libsqlite3-0 libsqlite3-dev libssl-dev libstdc++6 libtinfo-dev libtool libudev-dev libudev1 libuuid1 libvulkan-dev libvulkan1 libwayland-egl1-mesa libwww-perl libx11-6 libx11-xcb1 libxau6 libxcb1 libxcomposite1 libxcursor1 libxdamage1 libxdmcp6 libxext6 libxfixes3 libxi6 libxinerama1 libxkbcommon-dev libxrandr2 libxrender1 libxslt1-dev libxss-dev libxt-dev libxtst-dev libxtst6 linux-libc-dev-armhf-cross locales openbox p7zip patch perl php7.0-cgi pkg-config python python-cherrypy3 python-crypto python-dev python-numpy python-opencv python-openssl python-psutil python-yaml realpath rpm ruby subversion texinfo uuid-dev wdiff x11-utils xcompmgr xsltproc xutils-dev xvfb xz-utils zip zlib1g

It produced the following output:
Reading package lists...
Building dependency tree...
Reading state information...
fakeroot is already the newest version (1.20.2-1ubuntu1).
fakeroot set to manually installed.
libcap2 is already the newest version (1:2.24-12).
libffi-dev is already the newest version (3.2.1-4).
libffi-dev set to manually installed.
libffi6 is already the newest version (3.2.1-4).
libpcre3 is already the newest version (2:8.38-3.1).
libtinfo-dev is already the newest version (6.0+20160213-1ubuntu1).
libtinfo-dev set to manually installed.
libxau6 is already the newest version (1:1.0.8-1).
libxau6 set to manually installed.
libxcb1 is already the newest version (1.11.1-1ubuntu1).
libxcb1 set to manually installed.
libxdmcp6 is already the newest version (1:1.1.2-1.1).
libxdmcp6 set to manually installed.
libxext6 is already the newest version (2:1.3.3-1).
libxext6 set to manually installed.
xz-utils is already the newest version (5.1.1alpha+20120614-2ubuntu2).
xz-utils set to manually installed.
binutils is already the newest version (2.26.1-1ubuntu1~16.04.8).
binutils set to manually installed.
dpkg-dev is already the newest version (1.18.4ubuntu1.6).
dpkg-dev set to manually installed.
libc6 is already the newest version (2.23-0ubuntu11).
libc6-dev is already the newest version (2.23-0ubuntu11).
libc6-dev set to manually installed.
libglib2.0-0 is already the newest version (2.48.2-0ubuntu4.4).
libglib2.0-0 set to manually installed.
libx11-6 is already the newest version (2:1.6.3-1ubuntu2.1).
libx11-6 set to manually installed.
patch is already the newest version (2.7.5-1ubuntu0.16.04.2).
patch set to manually installed.
perl is already the newest version (5.22.1-9ubuntu0.6).
perl set to manually installed.
python is already the newest version (2.7.12-1~16.04).
python set to manually installed.
ruby is already the newest version (2:2.3.0+1bbox1~xenial1).
libstdc++6 is already the newest version (9.2.1-17ubuntu1~16.04).
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 binutils-mips64el-linux-gnuabi64 : Depends: binutils (= 2.26-8ubuntu2) but 2.26.1-1ubuntu1~16.04.8 is to be installed
 binutils-mipsel-linux-gnu : Depends: binutils (= 2.26-8ubuntu2) but 2.26.1-1ubuntu1~16.04.8 is to be installed
 elfutils : Depends: libdw1 (= 0.165-3ubuntu1) but 0.165-3ubuntu1.2 is to be installed
            Depends: libelf1 (= 0.165-3ubuntu1) but 0.165-3ubuntu1.2 is to be installed
 g++-5-multilib-arm-linux-gnueabihf : Depends: gcc-5-arm-linux-gnueabihf-base (= 5.3.1-14ubuntu2cross1) but 5.4.0-6ubuntu1~16.04.9cross1 is to be installed
                                      Depends: g++-5-arm-linux-gnueabihf (= 5.3.1-14ubuntu2cross1)
                                      Depends: libsfstdc++-5-dev-armhf-cross (= 5.3.1-14ubuntu2cross1) but it is not going to be installed or
                                               libsfstdc++-5-dev-armhf-cross but it is not going to be installed
 g++-arm-linux-gnueabihf : Depends: g++-5-arm-linux-gnueabihf (>= 5.3.1-3~)
 gcc-5-multilib-arm-linux-gnueabihf : Depends: gcc-5-arm-linux-gnueabihf-base (= 5.3.1-14ubuntu2cross1) but 5.4.0-6ubuntu1~16.04.9cross1 is to be installed
                                      Depends: gcc-5-arm-linux-gnueabihf (= 5.3.1-14ubuntu2cross1)
 gcc-arm-linux-gnueabihf : Depends: gcc-5-arm-linux-gnueabihf (>= 5.3.1-3~)

You will have to install the above packages yourself.
```

Reply:
Just guess, but a quick google of `sudo: unable to resolve host' gave this https://askubuntu.com/questions/59458/error-message-sudo-unable-to-resolve-host-none.

Maybe check in `/etc/hosts` that `holmes-nztrain` has an entry under localhost. On the train site there is no entry in `/etc/hosts` for `train.nzoi.org.nz`.

Reply:
I get the same error about `binutils-mips64el-linux-gnuabi64` dependencies, still investigating.

Reply:
Adding the hostname to `/etc/hosts` gets rid of the `sudo: unable to resolve host` message but the dependency errors still persist.

Reply:
It looks like debootstrap.bash creates an inconsistent /etc/apt/sources.list:
https://github.com/NZOI/nztrain/blob/42a9f42fad99acb1343cf1c71b7754e754474861/script/install/debootstrap.bash#L34-L40

Adding "universe" to `http://security.ubuntu.com/ubuntu $SUITE-security` fixes the problem.
I'll create a pull request to clean it up.

Reply:
I've tried building V8 inside chroot again with the above fix but ran into a few more issues:
- Python 2 needs to be installed otherwise `gclient` fails ([Python 2 is currently not being installed](https://github.com/NZOI/nztrain/pull/62))
```
+ gclient --version
/v8-build/depot_tools/gclient: line 32: exec: python: not found
```
- `fetch v8` fails. It seems that `gclient` isn't creating the required files. At some point I was able to get this working but I can't seem to reproduce the fix.
```
+ fetch v8
/v8-build/depot_tools/vpython: line 42: /v8-build/depot_tools/.cipd_bin/vpython: No such file or directory
```
- During the one time I managed to get the above working there was still a single unmet dependency when running `./build/install-build-deps.sh`, though the build still appeared to complete successfully so I'm unsure if it's actually needed.
```
The following packages have unmet dependencies:
 python-numpy : Depends: liblapack3 but it is not going to be installed or
                         liblapack.so.3

You will have to install the above packages yourself.
```

Reply:
So the issue with `fetch v8` seems to be fixable by moving `umount "$ISOLATE_ROOT/proc"` to after the V8 build under `debootstrap.bash`. Also note that `wget` needs to be installed else it gives the same error (`wget` is already installed under `debootstrap.bash` but I had it commented out while testing to make things faster).

Reply:
See 2488faa for a working build inside chroot. I didn't get the missing dependency this time but it's quite a bit slower inside chroot since there are more uninstalled build dependencies to download (takes me about 50 minutes to download+build). Also there are a bunch of `No such file or directory` errors/warnings ([see the build log here](https://pastebin.com/raw/1SpBzgBd)).

I'm not sure whether it'd be better to build V8 inside or outside chroot.

Reply:
Can you please try the following flags to reduce download size:
```diff
-git clone https://chromium.googlesource.com/chromium/tools/depot_tools.git
+git clone --depth 1 https://chromium.googlesource.com/chromium/tools/depot_tools.git

-fetch v8
+fetch --no-history v8

-gclient sync
+gclient sync --no-history

-./build/install-build-deps.sh
+./build/install-build-deps.sh --no-syms --no-arm --no-chromeos-fonts --no-nacl
```
(for me the last one reduced apt downloads in chroot from 400 MB to 244 MB).
If it's still too much we can build it outside chroot, but let's write the script in a way that makes it easier run in chroot later.

Regarding the locale warnings, `export LC_ALL=C` fixes them.

Reply:
I can confirm that this works (reduces the build to 26 minutes for me). I don't mind whether we build it inside or outside chroot.